src/object/tcb.c

```diff
@@ -257,80 +270,79 @@ tcb_queue_t tcbEPDequeue(tcb_t *tcb, tcb_queue_t queue)
 }
 
 #ifdef CONFIG_KERNEL_MCS
+
 void tcbReleaseRemove(tcb_t *tcb)
 {
     if (likely(thread_state_get_tcbInReleaseQueue(tcb->tcbState))) {
-        if (tcb->tcbSchedPrev) {
-            tcb->tcbSchedPrev->tcbSchedNext = tcb->tcbSchedNext;
-        } else {
-            NODE_STATE_ON_CORE(ksReleaseHead, tcb->tcbAffinity) = tcb->tcbSchedNext;
-            /* the head has changed, we might need to set a new timeout */
+        tcb_queue_t queue = NODE_STATE_ON_CORE(ksReleaseQueue, tcb->tcbAffinity);
+
+        if (queue.head == tcb) {
             NODE_STATE_ON_CORE(ksReprogram, tcb->tcbAffinity) = true;
         }
 
-        if (tcb->tcbSchedNext) {
-            tcb->tcbSchedNext->tcbSchedPrev = tcb->tcbSchedPrev;
-        }
+        NODE_STATE_ON_CORE(ksReleaseQueue, tcb->tcbAffinity) = tcb_queue_remove(queue, tcb);
 
-        tcb->tcbSchedNext = NULL;
-        tcb->tcbSchedPrev = NULL;
         thread_state_ptr_set_tcbInReleaseQueue(&tcb->tcbState, false);
     }
 }
 
+static inline ticks_t PURE tcbReadyTime(tcb_t *tcb)
+{
+    return refill_head(tcb->tcbSchedContext)->rTime;
+}
+
+static inline bool_t compare_times(word_t new_time, tcb_t *tcb)
+{
+    return tcb != NULL && new_time >= tcbReadyTime(tcb);
+}
+
+static tcb_t *compare_times_loop(word_t new_time, tcb_t *tcb)
+{
+    tcb_t *after = tcb;
+
+    while (compare_times(new_time, after)) {
+        after = after->tcbSchedNext;
+    }
+
+    return after;
+}
+
 void tcbReleaseEnqueue(tcb_t *tcb)
 {
     assert(thread_state_get_tcbInReleaseQueue(tcb->tcbState) == false);
     assert(thread_state_get_tcbQueued(tcb->tcbState) == false);
 
-    tcb_t *before = NULL;
-    tcb_t *after = NODE_STATE_ON_CORE(ksReleaseHead, tcb->tcbAffinity);
+    ticks_t new_time;
+    tcb_queue_t queue;
 
-    /* find our place in the ordered queue */
-    while (after != NULL &&
-           refill_head(tcb->tcbSchedContext)->rTime >= refill_head(after->tcbSchedContext)->rTime) {
-        before = after;
-        after = after->tcbSchedNext;
-    }
+    new_time = tcbReadyTime(tcb);
+    queue = NODE_STATE_ON_CORE(ksReleaseQueue, tcb->tcbAffinity);
 
-    if (before == NULL) {
-        /* insert at head */
-        NODE_STATE_ON_CORE(ksReleaseHead, tcb->tcbAffinity) = tcb;
+    if (!queue.head || new_time < tcbReadyTime(queue.head)) {
+        NODE_STATE_ON_CORE(ksReleaseQueue, tcb->tcbAffinity) = tcb_queue_prepend(queue, tcb);
         NODE_STATE_ON_CORE(ksReprogram, tcb->tcbAffinity) = true;
     } else {
-        before->tcbSchedNext = tcb;
-    }
-
-    if (after != NULL) {
-        after->tcbSchedPrev = tcb;
+        if (tcbReadyTime(queue.end) <= new_time) {
+            NODE_STATE_ON_CORE(ksReleaseQueue, tcb->tcbAffinity) = tcb_queue_append(queue, tcb);
+        } else {
+            tcb_t *after;
+            after = compare_times_loop(new_time, queue.head);
+            tcb_queue_insert(tcb, after);
+        }
```

Comment:
It might make it more readable, but that's not the reason they were taken out. The reason is that we need a name for them, so we can say what they correspond to on the spec level, which then allows us to split up the proof from one giant blob into multiple smaller pieces that are easier to handle.

Reply:
Fair enough. But in this case `compare_times` is very interwoven with `compare_times_loop`, so I'm sceptical it actually helps with the proof.